### DISCUSSION TITLE
posts/2024/03/how-to-reduce-size-in-git/

### DISCUSSION THREAD

**Discussion by giscus** [residualsun1/Residualsun#46] [General]
# posts/2024/03/how-to-reduce-size-in-git/

https://guozheng.rbind.io/posts/2024/03/how-to-reduce-size-in-git/

**Comment by CyrusYip** [comment]
我之前写了一篇《[GitHub 图床教程](https://cyrusyip.org/zh-cn/post/2020/12/05/host-images-on-github/)》，写得不太好。想着重写后再分享给你，没想到重写完你已经搞定了，动作好快！

补充一个将所有提交变成一个的方法：运行 `git rebase --interactive --root`，将第一行下面的提交都改成 fixup，保存并退出。

![image](https://github.com/residualsun1/Residualsun/assets/60951091/8ce4d5e1-0bf6-41d5-a034-2dd499868e6f)

**Reply by residualsun1** [reply]
> 我之前写了一篇《[GitHub 图床教程](https://cyrusyip.org/zh-cn/post/2020/12/05/host-images-on-github/)》，写得不太好。想着重写后再分享给你，没想到重写完你已经搞定了，动作好快！

教程很赞！其实想到可以节省仓库空间，我觉得这样的方式一点也不麻烦哈哈，而且要真影响少放点图片其实也蛮不错的，毕竟图片只是一种辅助，况且CDN的加载估计也要比在本地存放的加载快一些。

> 补充一个将所有提交变成一个的方法：运行`git rebase --interactive --root`，将第一行下面的提交都改成 fixup，保存并退出。

再过十个月后检查一下仓库大小，说不定我就会用上了（bushi

**Reply by CyrusYip** [reply]
@residualsun1 现在这个仓库应该不需要再压缩提交，纯文本文件很省空间的，我的博客仓库（包括主题）下载下来 `.git` 文件夹也就 6.2M。可能存静态文件的仓库用得上。

顺便说一下，最好不要修改公开仓库（比如开源项目）的历史，修改了历史别人再次拉取代码时会出错，别人也不知道你会不会加一些恶意代码。一般出了[把密码丢进公开仓库](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/removing-sensitive-data-from-a-repository#purging-a-file-from-your-repositorys-history)那么严重的事情才考虑修改历史。个人博客随意折腾，做正经的开源项目就要注意了。

**Reply by residualsun1** [reply]
> 顺便说一下，最好不要修改公开仓库（比如开源项目）的历史，修改了历史别人再次拉取代码时会出错，别人也不知道你会不会加一些恶意代码。一般出了[把密码丢进公开仓库](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/removing-sensitive-data-from-a-repository#purging-a-file-from-your-repositorys-history)那么严重的事情才考虑修改历史。个人博客随意折腾，做正经的开源项目就要注意了。

嗯，我这几天在网络上查教程的时候也经常看到不建议修改历史记录，即使是自己的博客也犹豫了好久哈哈。以后要用也只打算用在自己博客上了。

话说叶寻你是将这些说明文档都完整看过了吗？我发现你在统计之都论坛上就某个问题也会具体指引到某个文档的具体说明，感觉如果不是看过整个文档不会那么了解。

另，那现在如果你想帮忙修改我的博客，拉取代码的时候也会报错吗？

**Reply by CyrusYip** [reply]
> 嗯，我这几天在网络上查教程的时候也经常看到不建议修改历史记录，即使是自己的博客也犹豫了好久哈哈。以后要用也只打算用在自己博客上了。

你写文章的时候可以写一篇文章就建一个新分支，写完改好就合并到主分支。这样主分支就没那么多提交。或者建一个草稿分支，用 `git rebase` 压缩一些提交，再合并到主分支。又或者不要管你的提交数多不多。最好不要养成修改主分支历史的习惯。

---

> 话说叶寻你是将这些说明文档都完整看过了吗？我发现你在统计之都论坛上就某个问题也会具体指引到某个文档的具体说明，感觉如果不是看过整个文档不会那么了解。

参考文档（reference）不会完整看完，一般都是上谷歌搜到然后看需要的部分。看参考文档和查词典差不多，就是看自己需要的部分即可。如果你很有空，整个看也无妨。[文档可以分为4种](https://documentation.divio.com/)，根据自己需要选择。

![image](https://github.com/residualsun1/Residualsun/assets/60951091/5707c6b9-2e60-4333-a6e7-ee6797e2751c)

比如，一个新手要学习 Git，那就最好就看教程（比如 Pro Git）。要解决某个具体的问题就去看 Stack Overflow。

> 另，那现在如果你想帮忙修改我的博客，拉取代码的时候也会报错吗？

之前我就复刻了一份你的仓库在克隆到本地修改，现在会显示有冲突，在 GitHub  点 Sync fork 之后再本地 `git pull` 就会报错（fatal: Need to specify how to reconcile divergent branches.）。要执行 `git pull --force --rebase`。

<details>
  <summary>截图：</summary>

![image](https://github.com/residualsun1/Residualsun/assets/60951091/9053188b-be56-462a-a201-6512c2a5fdea)

![Screenshot_20240323_222832](https://github.com/residualsun1/Residualsun/assets/60951091/540b772f-fe48-4b75-84ba-31de002ba215)

</details>

**Reply by residualsun1** [reply]
> 你写文章的时候可以写一篇文章就建一个新分支，写完改好就合并到主分支。这样主分支就没那么多提交。或者建一个草稿分支，用 git rebase 压缩一些提交，再合并到主分支。又或者不要管你的提交数多不多。最好不要养成修改主分支历史的习惯。

哇！这是一个好思路！我可以试试。

> 参考文档（reference）不会完整看完，一般都是上谷歌搜到然后看需要的部分。看参考文档和查词典差不多，就是看自己需要的部分即可。如果你很有空，整个看也无妨。文档可以分为4种，根据自己需要选择。

嘿嘿，感谢指路~

> 之前我就复刻了一份你的仓库在克隆到本地修改，现在会显示有冲突，在 GitHub 点 Sync fork 之后再本地 git pull 就会报错（fatal: Need to specify how to reconcile divergent branches.）。要执行 git pull --force --rebase。

Git比我想象的还要复杂很多。只是看你写，我还是觉得`git rebase`有点复杂，等我找个时间再跟着GPT摸着石头过河吧！谢谢你，叶寻！

**Reply by CyrusYip** [reply]
提个建议，压缩提交之后可以把提交信息的标题写具体点。`Squashed commit of the following:` 这样就没啥有效信息，这是啥软件自动生成的吗？

**Reply by residualsun1** [reply]
@CyrusYip 接受建议:D。`Squashed commit of the following:`是我在GitHub Desktop中先commit草稿article分支，然后切换回主分支main之后，点击“Choose a branch to merge into main”，然后再选择“article”，最后点击“Squash and merge”。详细可见下方的视频。

[squash.webm](https://github.com/residualsun1/Residualsun/assets/130306032/d62e5ad4-4b38-4d49-8044-cb118d85cb55)
